lucet-runtime/lucet-runtime-internals/src/vmctx.rs

```diff
@@ -96,13 +99,24 @@ impl Vmctx {
     /// If the heap is already borrowed by `heap()` or `heap_mut()`, the instance will terminate
     /// with `TerminationDetails::BorrowError`.
     pub fn heap_mut(&self) -> RefMut<[u8]> {
+        unsafe {
+            self.reconstitute_heap_if_needed();
+        }
         let r = self
             .heap
             .try_borrow_mut()
             .unwrap_or_else(|_| panic!(TerminationDetails::BorrowError("heap_mut")));
         RefMut::map(r, |b| b.borrow_mut())
     }
 
+    unsafe fn reconstitute_heap_if_needed(&self) {
+        let inst = self.instance_mut();
+        if inst.heap_mut().len() != self.heap.borrow().len() {
+            let old_heap = self.heap.replace(Box::<[u8]>::from_raw(inst.heap_mut()));
+            Box::leak(old_heap);
```

Comment:
I'd appreciate a note here that this must be `Box::leak`'d to avoid `Box` trying to drop the heap, since all we're trying to do here is update the length of the boxed slice